File: nodns-cli/src/commands/resolve.rs
use clap::Args;
use nostr_sdk::nips::nip19::FromBech32;
use nostr_sdk::PublicKey;
use serde::Deserialize;

use crate::config::Config;
use crate::event;

const KIND_DNS: u16 = 11111;
const DEFAULT_API_BASE: &str = "https://nodns.shop";

#[derive(Args)]
pub struct CmdArgs {
    /// Domain name to resolve (e.g. npub1...nodns.shop)
    domain: String,

    /// Record type (A, AAAA, CNAME, TXT, MX, SRV, ANY)
    #[arg(short = 't', long, default_value = "ANY")]
    r#type: String,

    /// DNS server to query (host:port)
    #[arg(short, long)]
    server: Option<String>,

    /// Bot API base URL (default: https://nodns.shop)
    #[arg(long)]
    api_base: Option<String>,

    /// Query Nostr relays directly for raw events
    #[arg(long)]
    nostr: bool,

    /// Only query DNS, skip API and Nostr
    #[arg(long)]
    dns_only: bool,
}

#[derive(Deserialize)]
struct ApiRecordsResponse {
    records: Vec<ApiRecord>,
    count: usize,
}

#[derive(Deserialize)]
#[allow(dead_code)]
struct ApiRecord {
    npub: String,
    name: String,
    fqdn: String,
    #[serde(rename = "type")]
    record_type: String,
    ttl: u32,
    rdata: String,
    created_at: u64,
}

pub async fn run(args: CmdArgs, cfg: &Config) -> Result<(), String> {
    let filter_type = args.r#type.to_uppercase();

    if args.dns_only {
        println!(";; DNS ANSWER:");
        match resolve_dns(&args.domain, &filter_type, args.server.as_deref()).await {
            Ok(records) => {
                if records.is_empty() {
                    println!("  (no DNS records found)");
                }
                for record in records {
                    println!("  {record}");
                }
            }
            Err(e) => eprintln!("  (dns query failed: {e})"),
        }
        return Ok(());
    }

    if args.nostr {
        if let Some(npub_str) = extract_npub_from_domain(&args.domain) {
            println!(";; NOSTR EVENTS (kind {KIND_DNS}):");
            match resolve_nostr(cfg, &npub_str, &filter_type).await {
                Ok(events) => {
                    if events.is_empty() {
                        println!("  (no events found for {npub_str})");
                    }
                    for ev in &events {
                        for tag in ev.tags.iter() {
                            let vec = tag.clone().to_vec();
                            if vec.len() >= 5 && vec[0] == "record"
                                && (filter_type == "ANY" || vec[1].eq_ignore_ascii_case(&filter_type)) {
                                    let name = if vec[2].is_empty() { "@" } else { &vec[2] };
                                    println!("  {name}\t{}\t{}\t{}", vec[1], vec[3], vec[4]);
                                }
                        }
                    }
                }
                Err(e) => eprintln!("  (nostr query failed: {e})"),
            }
            println!();
        } else {
            eprintln!("Cannot extract npub from domain. Provide <npub>.nodns.shop domain.");
        }
        return Ok(());
    }

    let api_base = args.api_base.as_deref().unwrap_or(DEFAULT_API_BASE);
    let npub_str = extract_npub_from_domain(&args.domain);

    let api_result = if let Some(ref npub) = npub_str {
        resolve_api_by_npub(api_base, npub, &filter_type).await
    } else {
        resolve_api(api_base, &args.domain, &filter_type).await
    };

    println!(";; API RECORDS ({api_base}):");
    match api_result {
        Ok(resp) => {
            if resp.records.is_empty() {
                println!("  (no records found for {})", args.domain);
            }
            for r in &resp.records {
                if filter_type != "ANY" && !r.record_type.eq_ignore_ascii_case(&filter_type) {
                    continue;
                }
                let name = if r.name.is_empty() { "@" } else { &r.name };
                println!("  {name}\t{}\t{}\t{}\t{}", r.record_type, r.ttl, r.rdata, r.fqdn);
            }
            println!("  ;; {} record(s)", resp.count);
        }
        Err(e) => {
            eprintln!("  (api query failed: {e})");
            eprintln!("  ;; falling back to DNS...");
            match resolve_dns(&args.domain, &filter_type, args.server.as_deref()).await {
                Ok(records) => {
                    if records.is_empty() {
                        println!("  (no DNS records found)");
                    }
                    for record in records {
                        println!("  {record}");
                    }
                }
                Err(e2) => eprintln!("  (dns query also failed: {e2})"),
            }
        }
    }

    Ok(())
}

fn extract_npub_from_domain(domain: &str) -> Option<String> {
    let parts: Vec<&str> = domain.split('.').collect();
    if parts.len() >= 3 && parts[0].starts_with("npub1") {
        Some(parts[0].to_string())
    } else {
        None
    }
}

async fn resolve_api(
    api_base: &str,
    domain: &str,
    _filter_type: &str,
) -> Result<ApiRecordsResponse, String> {
    let url = format!("{api_base}/api/records?domain={domain}");
    let client = reqwest::Client::new();
    let resp = client
        .get(&url)
        .timeout(std::time::Duration::from_secs(10))
        .send()
        .await
        .map_err(|e| format!("{e}"))?;

    if !resp.status().is_success() {
        return Err(format!("HTTP {}", resp.status()));
    }

    resp.json::<ApiRecordsResponse>()
        .await
        .map_err(|e| format!("parse error: {e}"))
}

async fn resolve_api_by_npub(
    api_base: &str,
    npub: &str,
    _filter_type: &str,
) -> Result<ApiRecordsResponse, String> {
    let url = format!("{api_base}/api/records/by-npub/{npub}");
    let client = reqwest::Client::new();
    let resp = client
        .get(&url)
        .timeout(std::time::Duration::from_secs(10))
        .send()
        .await
        .map_err(|e| format!("{e}"))?;

    if !resp.status().is_success() {
        return Err(format!("HTTP {}", resp.status()));
    }

    resp.json::<ApiRecordsResponse>()
        .await
        .map_err(|e| format!("parse error: {e}"))
}

async fn resolve_nostr(
    cfg: &Config,
    npub: &str,
    filter_type: &str,
) -> Result<Vec<nostr_sdk::Event>, String> {
    let pk = if npub.starts_with("npub1") {
        PublicKey::from_bech32(npub).map_err(|e| format!("invalid npub: {e}"))?
    } else {
        PublicKey::from_hex(npub).map_err(|e| format!("invalid pubkey: {e}"))?
    };

    let _ = filter_type;
    event::fetch_events(cfg, Some(pk)).await
}

async fn resolve_dns(
    domain: &str,
    record_type_str: &str,
    server: Option<&str>,
) -> Result<Vec<String>, String> {
    use hickory_proto::rr::RecordType;
    use std::str::FromStr;

    let record_type = match record_type_str {
        "A" => RecordType::A,
        "AAAA" => RecordType::AAAA,
        "CNAME" => RecordType::CNAME,
        "TXT" => RecordType::TXT,
        "MX" => RecordType::MX,
        "SRV" => RecordType::SRV,
        "ANY" => RecordType::ANY,
        other => return Err(format!("unsupported record type: {other}")),
    };

    let name = hickory_proto::rr::Name::from_str(domain)
        .map_err(|e| format!("invalid domain: {e}"))?;

    let resolver = if let Some(_srv) = server {
        hickory_resolver::TokioResolver::builder_tokio()
            .map_err(|e| format!("resolver init failed: {e}"))?
            .build()
    } else {
        hickory_resolver::TokioResolver::builder_tokio()
            .map_err(|e| format!("resolver init failed: {e}"))?
            .build()
    };

    let response = resolver
        .lookup(name, record_type)
        .await
        .map_err(|e| format!("query failed: {e}"))?;

    Ok(response.records().iter().map(|r| format!("{r}")).collect())
}

File: nodns-cli/src/commands/resolve.ts
import { Command } from "commander";
import * as dns from "node:dns/promises";
import { decode as nip19Decode } from "nostr-tools/nip19";
import { fetchEvents } from "../lib/nostr.js";
import { DEFAULT_API_BASE, readRelays } from "../lib/constants.js";
import type { DnsRecord } from "../lib/types.js";

function extractNpubFromDomain(domain: string): string | null {
  const parts = domain.split(".");
  if (parts.length >= 3 && parts[0].startsWith("npub1")) {
    return parts[0];
  }
  return null;
}

function extractRecords(tags: string[][]): DnsRecord[] {
  const records: DnsRecord[] = [];
  for (const tag of tags) {
    if (tag.length >= 5 && tag[0] === "record") {
      records.push({
        type: tag[1],
        name: tag[2],
        ttl: parseInt(tag[3], 10) || 3600,
        rdata: tag[4],
      });
    }
  }
  return records;
}

function pubkeyNpubToHex(npub: string): string | null {
  try {
    const decoded = nip19Decode(npub);
    if (decoded.type !== "npub") return null;
    return decoded.data;
  } catch {
    return null;
  }
}

async function resolveViaDns(
  domain: string,
  recordType: string,
): Promise<void> {
  console.log(";; DNS ANSWER:");
  try {
    const type = recordType.toUpperCase();
    if (type === "A" || type === "ANY") {
      const addrs = await dns.resolve4(domain).catch(() => []);
      for (const a of addrs) console.log(`  ${domain}.\t\tA\t${a}`);
    }
    if (type === "AAAA" || type === "ANY") {
      const addrs = await dns.resolve6(domain).catch(() => []);
      for (const aaaa of addrs) console.log(`  ${domain}.\t\tAAAA\t${aaaa}`);
    }
    if (type === "CNAME" || type === "ANY") {
      const cnames = await dns.resolveCname(domain).catch(() => []);
      for (const c of cnames) console.log(`  ${domain}.\t\tCNAME\t${c}`);
    }
    if (type === "TXT" || type === "ANY") {
      const txts = await dns.resolveTxt(domain).catch(() => []);
      for (const t of txts) console.log(`  ${domain}.\t\tTXT\t"${t.join("")}"`);
    }
    if (type === "MX" || type === "ANY") {
      const mxs = await dns.resolveMx(domain).catch(() => []);
      for (const mx of mxs) console.log(`  ${domain}.\t\tMX\t${mx.priority} ${mx.exchange}`);
    }
  } catch (e) {
    console.error(`  (dns query failed: ${e instanceof Error ? e.message : String(e)})`);
  }
}

async function resolveViaNostr(
  relays: string[],
  npubStr: string,
): Promise<void> {
  console.log(";; NOSTR EVENTS (kind 11111):");
  const hex = pubkeyNpubToHex(npubStr);
  if (!hex) {
    console.error("  (invalid npub)");
    return;
  }

  try {
    const events = await fetchEvents(relays, hex);
    if (events.length === 0) {
      console.log(`  (no events found for ${npubStr})`);
      return;
    }
    for (const ev of events) {
      const records = extractRecords(ev.tags);
      for (const r of records) {
        const name = r.name || "@";
        console.log(`  ${name}\t${r.type}\t${r.ttl}\t${r.rdata}`);
      }
    }
  } catch (e) {
    console.error(`  (nostr query failed: ${e instanceof Error ? e.message : String(e)})`);
  }
}

interface ApiRecord {
  name: string;
  fqdn: string;
  type: string;
  ttl: number;
  rdata: string;
}

async function resolveViaApi(
  apiBase: string,
  domain: string,
  npubStr: string | null,
): Promise<void> {
  console.log(`;; API RECORDS (${apiBase}):`);
  try {
    let url: string;
    if (npubStr) {
      url = `${apiBase}/api/records/by-npub/${npubStr}`;
    } else {
      url = `${apiBase}/api/records?domain=${encodeURIComponent(domain)}`;
    }
    const res = await fetch(url, { signal: AbortSignal.timeout(10_000) });
    if (!res.ok) {
      console.error(`  (api query failed: HTTP ${res.status})`);
      console.error("  ;; falling back to DNS...");
      await resolveViaDns(domain, "ANY");
      return;
    }
    const data = (await res.json()) as { records?: ApiRecord[]; count?: number };
    const records = data.records ?? [];
    if (records.length === 0) {
      console.log(`  (no records found for ${domain})`);
      return;
    }
    for (const r of records) {
      const name = r.name || "@";
      console.log(`  ${name}\t${r.type}\t${r.ttl}\t${r.rdata}\t${r.fqdn}`);
    }
    console.log(`  ;; ${data.count ?? records.length} record(s)`);
  } catch (e) {
    console.error(`  (api query failed: ${e instanceof Error ? e.message : String(e)})`);
    console.error("  ;; falling back to DNS...");
    await resolveViaDns(domain, "ANY");
  }
}

export const resolveCommand = new Command("resolve")
  .description("Resolve a DNS name")
  .argument("<domain>", "Domain name to resolve")
  .option("-t, --type <type>", "Record type", "ANY")
  .option("--dns", "Query DNS only")
  .option("--nostr", "Query Nostr relays for raw events")
  .option("--api-base <url>", "Bot API base URL", DEFAULT_API_BASE)
  .action(async (domain: string, opts, cmd: Command) => {
    const o = cmd.optsWithGlobals();
    const relays = readRelays(o.relay);
    const recordType = (opts.type as string).toUpperCase();
    const npubStr = extractNpubFromDomain(domain);

    if (opts.dns as boolean) {
      await resolveViaDns(domain, recordType);
      return;
    }

    if (opts.nostr as boolean) {
      if (!npubStr) {
        console.error("Cannot extract npub from domain. Use <npub>.nodns.shop format.");
        return;
      }
      await resolveViaNostr(relays, npubStr);
      console.log();
      return;
    }

    await resolveViaApi(opts.apiBase as string, domain, npubStr);
  });
